### PR TITLE
make sure domains actually get right params

### PR DESCRIPTION
**Pull request by hbbtstar**
We were running into an issue where get_all_query_params_by_domain() was not returning the right params for expandable urls (I.E. it would not return 'q' for google.ca when the entry in search_engines.json was google.{}). This pull request expands the test coverage for lots of these domains, as well as dynamically expands those domains in _get_search_engines.

I thought maybe it would be faster to pre-compute these domain expansions and output a json file with all the engines, but surprisingly that was not the case: dynamically expanding in-file turned out to be slightly faster.

Unfortunately, tldextract.extract is really slow, and this ballooned the time it took to run get_all_query_params_by_domain() considerably since it got called for every now-expanded domain in the list. I was able to cut down the time considerably by using a regex instead of tldextract; since we have a clearly delineated list of possible url structures in the search_engines.json, I felt we could test adequately against miscategorization. I am very aware of the fallibility of homegrown regex vs tldextract, however, so if the much slower runtime of get_all_query_params_by_domain() is acceptable we can stick with tldextract ( as it stands, all the edge case tests pass). Here is a gist of some basic timeit runs:

https://gist.github.com/hbbtstar/e37415c457db023efef0e5f1b1f5f53e

(the current 0.6.0 implementation of get_all_query_params_by_domain() is way faster than any other, but it's also broken, so that's artificially low)

@dan-blanchard, I would appreciate some more eyes on this: you may be able to spot something here I missed re: making this more performant. As it stands, extract() is as performant as it was before, so I'm not too worried there, but going from 0.01 seconds to 0.05 seconds per run of getting all the params is a larger change.

### CODE REVIEW

**Comment by hbbtstar** [comment]
on further review, there are too many edge cases here: the tldextract did come back to bite me. Will keep poking

File: serpextract/serpextract.py [hunk 1]
```diff
@@ -193,6 +196,20 @@ def _get_search_engines():
...
+    urls = urls if isinstance(urls, list) else [urls]
+    urls = set(urls)
+    end_string = re.compile('\w$')
+    expanded_urls = {url.format(country_code) for url in urls
+                     for country_code in _country_codes}
+    expanded_urls.update({url.format('{}.{}'.format(second_level_domain, cc_sub_domain))
+                          for url in urls
+                          for cc_sub_domain in _country_codes
+                          for second_level_domain in second_level_domains
+                          if not end_string.search(url)})
```
Comment (dan-blanchard) [inline serpextract/serpextract.py]: `url[-1].isalnum()` is going to be much faster than using a regex, so you can get rid of this regex (and the compilation of it) entirely.

File: serpextract/serpextract.py [hunk 2]
```diff
@@ -506,23 +490,26 @@ def get_all_query_params():
 
     return list(all_params)
 
-
+qs_params = defaultdict(list)
```
Comment (dan-blanchard) [inline serpextract/serpextract.py]: This should start with `_`, we don't want this cache to be considered part of the public API.

File: serpextract/serpextract.py [hunk 3]
```diff
@@ -506,23 +490,26 @@ def get_all_query_params():
...
+    global qs_params
+    if qs_params:
+        return qs_params
     engines = _get_search_engines()
     param_dict = defaultdict(list)
     for domain, parser in iteritems(engines):
-        # Find non-regex params
         params = {param for param in parser.keyword_extractor
                   if isinstance(param, string_types)}
-        tld_res = tldextract.extract(domain)
```
Comment (dan-blanchard) [inline serpextract/serpextract.py]: Nitpick: Since you didn't actually change anything here, please change the name back to what it was and add the comment back.

File: serpextract/serpextract.py [hunk 4]
```diff
@@ -558,11 +544,7 @@ def get_parser(referring_url):
     # search engines, yahoo and yahoo images
     if u'{}{}'.format(domain, path) in engines:
         engine_key = u'{}{}'.format(domain, path)
-    elif u'{}{}'.format(lossy_domain, path) in engines:
-        engine_key = u'{}{}'.format(lossy_domain, path)
-    elif lossy_domain in engines:
-        engine_key = lossy_domain
-    elif domain not in engines:
+    elif domain not in engines and domain.strip('www.') not in engines:
```
Comment (dan-blanchard) [inline serpextract/serpextract.py]: Ah, this is one of my favorite super common bugs.

```py
In [21]: domain = 'www.wegotbugs.com'

In [22]: domain.strip('www.')
Out[22]: 'egotbugs.com'
```

You want `domain[4:] if domain.startswith('www.') else domain`.

File: serpextract/serpextract.py [hunk 5]
```diff
@@ -578,7 +560,7 @@ def get_parser(referring_url):
         else:
             return None
 
-    return engines.get(engine_key)
+    return engines.get(engine_key) or engines.get(engine_key.strip('www.'))
```
Comment (dan-blanchard) [inline serpextract/serpextract.py]: Same bug here.
